Custom Middleware Reusability

## Expected Behavior
Defining a custom middleware once should allow to reuse it in `customMiddleware` more than once. This might be achieved by giving middlewares an id (to reference custom middlewares) and a name (used for order and display).

## Current Behavior
It fails with `duplicate middleware name`:
```
Error: Failed to add custom middleware proxy. A middleware with the same name is already known.
    at MiddlewareManager.addCustomMiddleware (/XXX/node_modules/@ui5/cli/node_modules/@ui5/server/lib/middleware/MiddlewareManager.js:159:11)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! XXX@1.0.0 start: `ui5 serve -p 3000`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the XXX@1.0.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

## Steps to reproduce the issue
1. Create a simple custom middleware
2. Register as custom middleware in ui5.yaml
3. Use the newly registered middleware twice in ui5.yaml

## Affected components *(if known)*
- [ ] [ui5-builder](https://github.com/SAP/ui5-builder)
- [X] [ui5-server](https://github.com/SAP/ui5-server)
- [ ] [ui5-cli](https://github.com/SAP/ui5-cli)
- [ ] [ui5-fs](https://github.com/SAP/ui5-fs)
- [ ] [ui5-project](https://github.com/SAP/ui5-project)
- [ ] [ui5-logger](https://github.com/SAP/ui5-logger)

## Log Output / Stack Trace
```
Error: Failed to add custom middleware proxy. A middleware with the same name is already known.
    at MiddlewareManager.addCustomMiddleware (/XXX/node_modules/@ui5/cli/node_modules/@ui5/server/lib/middleware/MiddlewareManager.js:159:11)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! XXX@1.0.0 start: `ui5 serve -p 3000`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the XXX@1.0.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

--- a/lib/middleware/MiddlewareManager.js
+++ b/lib/middleware/MiddlewareManager.js
@@ -29,15 +29,26 @@ class MiddlewareManager {
 		});
 	}
 
-	async addMiddleware(middlewareName, {
+	async addMiddleware(configuredMiddlewareName, {
 		wrapperCallback, mountPath = "/",
 		beforeMiddleware, afterMiddleware
 	} = {}) {
-		let middlewareCallback = middlewareRepository.getMiddleware(middlewareName);
+		let middlewareCallback = middlewareRepository.getMiddleware(configuredMiddlewareName);
 		if (wrapperCallback) {
 			middlewareCallback = wrapperCallback(middlewareCallback);
 		}
-		if (this.middleware[middlewareName] || this.middlewareExecutionOrder.includes(middlewareName)) {
+
+		let middlewareName = configuredMiddlewareName;
+		if (this.middleware[middlewareName]) {
+			// Middleware is already known
+			// => add a suffix to allow for multiple configurations of the same middleware
+			let suffixCounter = 0;
+			while (this.middleware[middlewareName]) {
+				suffixCounter++; // Start at 1
+				middlewareName = `${configuredMiddlewareName}--${suffixCounter}`;
+			}
+		}
+		if (this.middlewareExecutionOrder.includes(middlewareName)) {
 			throw new Error(`Failed to add duplicate middleware ${middlewareName}`);
 		}
 
@@ -154,17 +165,7 @@ class MiddlewareManager {
 					`defines neither a "beforeMiddleware" nor an "afterMiddleware" parameter. One must be defined.`);
 			}
 
-			let middlewareName = middlewareDef.name;
-			if (this.middleware[middlewareName]) {
-				// Middleware is already known
-				// => add a suffix to allow for multiple configurations of the same middleware
-				let suffixCounter = 0;
-				while (this.middleware[middlewareName]) {
-					suffixCounter++; // Start at 1
-					middlewareName = `${middlewareDef.name}--${suffixCounter}`;
-				}
-			}
-			await this.addMiddleware(middlewareName, {
+			await this.addMiddleware(middlewareDef.name, {
 				wrapperCallback: (middleware) => {
 					return ({resources}) => {
 						const options = {

--- a/lib/middleware/MiddlewareManager.js
+++ b/lib/middleware/MiddlewareManager.js
@@ -154,12 +154,17 @@ class MiddlewareManager {
 					`defines neither a "beforeMiddleware" nor an "afterMiddleware" parameter. One must be defined.`);
 			}
 
-			if (this.middleware[middlewareDef.name]) {
+			let middlewareName = middlewareDef.name;
+			if (this.middleware[middlewareName]) {
 				// Middleware is already known
-				throw new Error(`Failed to add custom middleware ${middlewareDef.name}. ` +
-					`A middleware with the same name is already known.`);
+				// => add a suffix to allow for multiple configurations of the same middleware
+				let suffixCounter = 0;
+				while (this.middleware[middlewareName]) {
+					suffixCounter++; // Start at 1
+					middlewareName = `${middlewareDef.name}--${suffixCounter}`;
+				}
 			}
-			await this.addMiddleware(middlewareDef.name, {
+			await this.addMiddleware(middlewareName, {
 				wrapperCallback: (middleware) => {
 					return ({resources}) => {
 						const options = {

--- a/lib/middleware/MiddlewareManager.js
+++ b/lib/middleware/MiddlewareManager.js
@@ -49,7 +49,7 @@ class MiddlewareManager {
 			}
 		}
 		if (this.middlewareExecutionOrder.includes(middlewareName)) {
-			throw new Error(`Failed to add duplicate middleware ${middlewareName}`);
+			throw new Error(`Middleware ${middlewareName} already added to execution order. This should not happen.`);
 		}
 
 		if (beforeMiddleware || afterMiddleware) {

--- a/test/lib/server/middleware/MiddlewareManager.js
+++ b/test/lib/server/middleware/MiddlewareManager.js
@@ -86,6 +86,32 @@ test("addMiddleware: Adding already added middleware produces unique middleware 
 	t.truthy(middlewareManager.middleware["serveIndex--2"], "Middleware got added to internal map with unique name");
 	t.deepEqual(middlewareManager.middleware["serveIndex--2"].mountPath, "/goose",
 		"Middleware got added correct mount path");
+
+	t.deepEqual(middlewareManager.middlewareExecutionOrder, [
+		"serveIndex",
+		"serveIndex--1",
+		"serveIndex--2"
+	], "Middlewares got added to middlewareExecutionOrder in correct order and with correct unique names");
+});
+
+test("addMiddleware: Adding middleware already added to middlewareExecutionOrder", async (t) => {
+	const middlewareManager = new MiddlewareManager({
+		tree: {},
+		resources: {
+			all: "I",
+			rootProject: "like",
+			dependencies: "ponies"
+		}
+	});
+
+	middlewareManager.middlewareExecutionOrder.push("serveIndex");
+
+	const err = await t.throwsAsync(() => {
+		return middlewareManager.addMiddleware("serveIndex");
+	});
+	t.deepEqual(err.message,
+		"Middleware serveIndex already added to execution order. This should not happen.",
+		"Rejected with correct error message");
 });
 
 test("addMiddleware: Add middleware", async (t) => {

--- a/test/lib/server/middleware/MiddlewareManager.js
+++ b/test/lib/server/middleware/MiddlewareManager.js
@@ -58,7 +58,7 @@ test("applyMiddleware", async (t) => {
 	t.deepEqual(appUseStub.getCall(0).args[1], "myMiddleware", "app.use got called with correct middleware parameter");
 });
 
-test("addMiddleware: Add already added middleware", async (t) => {
+test("addMiddleware: Adding already added middleware produces unique middleware name", async (t) => {
 	const middlewareManager = new MiddlewareManager({
 		tree: {},
 		resources: {
@@ -68,11 +68,24 @@ test("addMiddleware: Add already added middleware", async (t) => {
 		}
 	});
 
-	await middlewareManager.addMiddleware("serveIndex");
-	const err = await t.throwsAsync(() => {
-		return middlewareManager.addMiddleware("serveIndex");
+	await middlewareManager.addMiddleware("serveIndex", {
+		mountPath: "/pony"
 	});
-	t.deepEqual(err.message, "Failed to add duplicate middleware serveIndex", "Rejected with correct error message");
+	await middlewareManager.addMiddleware("serveIndex", {
+		mountPath: "/seagull"
+	});
+	await middlewareManager.addMiddleware("serveIndex", {
+		mountPath: "/goose"
+	});
+	t.truthy(middlewareManager.middleware["serveIndex"], "Middleware got added to internal map with unique name");
+	t.deepEqual(middlewareManager.middleware["serveIndex"].mountPath, "/pony",
+		"Middleware got added correct mount path");
+	t.truthy(middlewareManager.middleware["serveIndex--1"], "Middleware got added to internal map with unique name");
+	t.deepEqual(middlewareManager.middleware["serveIndex--1"].mountPath, "/seagull",
+		"Middleware got added correct mount path");
+	t.truthy(middlewareManager.middleware["serveIndex--2"], "Middleware got added to internal map with unique name");
+	t.deepEqual(middlewareManager.middleware["serveIndex--2"].mountPath, "/goose",
+		"Middleware got added correct mount path");
 });
 
 test("addMiddleware: Add middleware", async (t) => {
@@ -330,7 +343,7 @@ test("addCustomMiddleware: Custom middleware got added", async (t) => {
 		"addMiddleware was called with correct afterMiddleware option");
 });
 
-test("addCustomMiddleware: Custom middleware with duplicate name", async (t) => {
+test("addCustomMiddleware: No special handling for custom middleware with duplicate name", async (t) => {
 	const project = {
 		metadata: {
 			name: "my project"
@@ -351,12 +364,11 @@ test("addCustomMiddleware: Custom middleware with duplicate name", async (t) => 
 		}
 	});
 	middlewareManager.middleware["my custom middleware A"] = true;
-	middlewareManager.middleware["my custom middleware A--1"] = true;
 	const addMiddlewareStub = sinon.stub(middlewareManager, "addMiddleware").resolves();
 	await middlewareManager.addCustomMiddleware();
 
 	t.deepEqual(addMiddlewareStub.callCount, 1, "addMiddleware was called once");
-	t.deepEqual(addMiddlewareStub.getCall(0).args[0], "my custom middleware A--2",
+	t.deepEqual(addMiddlewareStub.getCall(0).args[0], "my custom middleware A",
 		"addMiddleware was called with correct middleware name");
 });
 

--- a/test/lib/server/middleware/MiddlewareManager.js
+++ b/test/lib/server/middleware/MiddlewareManager.js
@@ -338,7 +338,7 @@ test("addCustomMiddleware: Custom middleware with duplicate name", async (t) => 
 		server: {
 			customMiddleware: [{
 				name: "my custom middleware A",
-				afterMiddleware: "my custom middleware A"
+				afterMiddleware: "serveIndex"
 			}]
 		}
 	};
@@ -351,15 +351,13 @@ test("addCustomMiddleware: Custom middleware with duplicate name", async (t) => 
 		}
 	});
 	middlewareManager.middleware["my custom middleware A"] = true;
+	middlewareManager.middleware["my custom middleware A--1"] = true;
 	const addMiddlewareStub = sinon.stub(middlewareManager, "addMiddleware").resolves();
-	const err = await t.throwsAsync(() => {
-		return middlewareManager.addCustomMiddleware();
-	});
+	await middlewareManager.addCustomMiddleware();
 
-	t.deepEqual(err.message, "Failed to add custom middleware my custom middleware A. " +
-		"A middleware with the same name is already known.",
-	"Rejected with correct error message");
-	t.deepEqual(addMiddlewareStub.callCount, 0, "Add middleware did not get called");
+	t.deepEqual(addMiddlewareStub.callCount, 1, "addMiddleware was called once");
+	t.deepEqual(addMiddlewareStub.getCall(0).args[0], "my custom middleware A--2",
+		"addMiddleware was called with correct middleware name");
 });
 
 test("addCustomMiddleware: Missing name configuration", async (t) => {

